MAINT: interpolate: trim down `_bspl` extension

ev-br opened the pull request:
This is a work-in-progress attempt at trimming down the binary size of scipy.interpolate. Previously, https://github.com/scipy/scipy/pull/21539 and https://github.com/scipy/scipy/pull/21223 managed to shave off some 160kb by removing fused memoryview types from a Cython extension and moving things off Cython to pure python.  

This PR aims to see the effect of ditching Cython completely and replacing it with a manually written C extension. 
It is a little too early to conclusively assess the full effects on the binary size: there are three heavy functions left in the Cython `_bspl` extension.
It might be good time though to start assessing the downsides on clarity and maintainability: Cython's much more expressive and less error-prone than a manual C extension. Or maybe it's just me being used to Cython :-).

The design here is trying to separate the workhorse routines (`interpolate/src/__fitpack.{h,cc}`) and python-facing wrappers (`interpolate/src/_dierckxmodule.cc`). This way, computational routines know nothing about python, and the glue does not do any computations other than parsing python arguments, doing sanity checking, and sending inputs to computational routines. I'm also keeping all memory allocations in the glue layer. 

So at this stage, I'd appreciate any comments about the C API --- does the general picture make sense? is the code readable? does it look maintainable? I'm sure there are multiple problems and bugs, too.

ev-br committed:
MAINT: interpolate: move evaluate_spline to C++

ev-br committed:
MAINT: interpolate: move fpknot wrap from Cython to manual

ev-br committed:
MAINT: interpolate" move fpback wrap from Cython to manual

ev-br committed:
MAINT: interpolate: move qr_reduce wrap from Cython to manual

ev-br committed:
simplify fpknot

ev-br committed:
MAINT: catch C++ exceptions, address Nathan's comments

ev-br committed:
MAINT: interpolate: move _colloc to C++

ev-br committed:
MAINT: interpolate: move _norm_eq_lsq to C++

ev-br committed:
MAINT: move py_find_interval from Cython to manual

ev-br committed:
MAINT: interpolate: move _coloc wrap from Cython to manual

ev-br committed:
BUG: interpolate: do not leak wrk arrays

ev-br committed:
MAINT: interpolate: move norm_eq_lsq from Cython to manual

ev-br committed:
MAINT: interpolate: move evaluate_spline from Cython to manual

ev-br committed:
MAINT: interpolate: add docstrings to _dierckxmodule from _bspl.pyx

ev-br committed:
MAINT: interpolate: rm dead/commented out code from _bspl.pyx

ev-br committed:
BUG: interpolate: fix a function name

From python, it's find_interval, not py_find_interval

ev-br committed:
MAINT: interpolate: rm find_interval cython wrapper

ev-br committed:
MAINT: interpolate: move evaluate_all_bspl from cython to manual

ev-br committed:
BLD: interpolate: hide extra symbols in _dierckxmodule.so

Otherwise, 32-bit build on linux in CI fails on
./tools/check_pyext_symbol_hiding.sh

ev-br committed:
BUG: interpolate: make _dierckx indexing int64_t throughout

ev-br committed:
MAINT: interpolate: appease mypy

ilayn commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -290,18 +292,18 @@ fpback( /* inputs*/
-    for (ssize_t i=nc-2; i >= 0; --i) {
-        ssize_t nel = std::min(nz, nc - i);
-        for (ssize_t l=0; l < ydim2; ++l){
+    for (d_ssize_t i=nc-2; i >= 0; --i) {
```
It's very minor thing but if you define all variables at the top Fortran style, it becomes much easier for the compiler to detect unused vars and also reduces the mistyped vars acting stupid with implicit casts.

ilayn commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -418,9 +420,152 @@ fpknot(const double *x_ptr, ssize_t m,
+*/
+void
+_evaluate_spline(
+    const double *tptr, d_ssize_t len_t,         // t, shape (len_t,)
```
if these d_ssize_t vars are not changing add `const` to them so compiler can do more magic with them.

ev-br replied:
Gotta love const-correctness :-). Given that these size variables come from `PyArray_DIMS(array, 0)`, which is an `npy_intp` not `const npy_intp`, let me see what sort of hoops I'll need to jump through in the wrapper. Once I figure out the `npy_intp` vs `ptrdiff_t` integer type madness across 32-bit linux, windows and whatever clang is doing on MacOS.

I think I gave up trying to be "idiomatic" with `ptrdiff_t` on the C++ side, `npy_intp` on the Numpy C API side and `dtype=np.intp` in python. The current attempt is to use `int64_t` for all indexing (which is what `d_ssize_t` is typedef-d to), let's see what the CI battery is to say about it. When it deigns to show a green checkmark, I'll see about const-correctness.

ilayn replied:
I'm not comment on C++ things. You dug your own hole you deal with it 😅

ev-br replied:
Nah, that's exactly the same in C, tracing back to K&R : int vs long vs long long types being platform-dependent. 

I _am_ digging my own hole in a different place: I want a clear separation between C (C++, doesn't matter) and the C-to-python glue. This means `__fitpack.{h,cc}` does not  know about `Py_ssize_t` or `npy_intp` types. Then the need to pass an integer array between python and C via numpy C API descends into madness even if I control all of it.

I _suspect_ it will be a bit saner when we only have numpy 2.0, but for now it's... uhm... iterative. 

On a related subject: do you see downsides of using `int64_t` for all indexing? (Size of arrays is not a concern, just buy some more memory, sorry).

ilayn replied:
As long as you don't do BLAS/LAPACK calls I don't think so, it should be good to go.

lorentzenchr replied:
I think it is perfectly fine to use elementary types like `Py_ssize_t` or `npy_intp` in the "non-glue" C/C++ part.

ilayn commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -418,9 +420,152 @@ fpknot(const double *x_ptr, ssize_t m,
+
+            // Form linear combinations
+            for (d_ssize_t jp=0; jp < m; jp++) {
+                out(ip, jp) = 0.0;
```
Maybe 0.0 initialize out outside the loop so this is more prefetchable?

ev-br replied:
Hm. So you're saying that a separate loop zeroing out the buffer is preferable because of... what's more prefetchable?

ilayn replied:
Meaning that the loop does not interrupt. And also faster to zero initialize. But I guess these are micro optimizations that are not worth it. Yeah, probably just nitpicks come to think of it.

ev-br replied:
That's actually interesting :-). I'll try to do some time measurements in a bit--- will be happy when this graduates to where there kinds of optimizations are the main concern. Thank you!

ilayn commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -418,9 +420,152 @@ fpknot(const double *x_ptr, ssize_t m,
+void
+_coloc_matrix(const double *xptr, d_ssize_t m,       // x, shape(m,)
+              const double *tptr, d_ssize_t len_t,   // t, shape(len_t,)
+              int k,
```
same `const` comment. Also it makes it obvious that they are input parameters.

ev-br committed:
MAINT: interpolate/__fitpack: move all LAPACK stuff close together

ev-br committed:
MAINT: interpolate: remove Cython implementation of BSpline.design_matrix

Reuse _dierckx.data_matrix, which constructs the same object in a "packed" format,
with per-row offsets of non-zero elements instead of per-element CSR indices/indptr
arrays.

Add `extrapolate=False` parameter to data_matrix to account for design_marix usages.

ev-br commented:
OK, some progress:

- C int madness is fixed.
- `_bspl.pyx` Cython extension only has only two functions left, `evaluate_ndbspline` and `make_nd`, both NdBSpline related.
- Removing `make_design_matrix` and its fused int32/int64 dtype gives back another 60kb in the binary size, so that the total binary size locally is 300Kb from Cython and 64Kb from C++. (compare to 612Kb pre gh-21314).

Heads up @lorentzenchr, this refactors `BSpline.design_matrix` which you worked on for scikit-learn IIRC.

lorentzenchr reviewed (commented):
Some minor comments.

ilayn commented on scipy/interpolate/src/__fitpack.h:
```diff
@@ -1,31 +1,58 @@
+// use int64_t for indexing throughout
+// Using system dependent int types (ssize_t, ptrdiff_t, Py_ssize_t and npy_intp)
+// is a world of pain on all of Windows; 32-bit & 64-bit linux & numpy 1.x.
+typedef int64_t d_ssize_t;
```
Doesn't C++ have `<stdint.h>`?

ev-br replied:
Yes, why? This is where `int64_t` declaration comes from. (Technically, it's `#include <cinttypes>` in this header, but that's the same I believe).

ilayn replied:
Then you can use it regardless of the platform without typedefing like @lorentzenchr mentioned, with guarantee that platform will give you 32/64bit int.

ev-br replied:
Ok, done. As of the last commit, the `d_ssize_t` typedef is gone, all indexing uses an explicit int64_t

ev-br commented on scipy/interpolate/src/_dierckxmodule.cc:
```diff
@@ -0,0 +1,774 @@
+{
+    PyObject *py_x=NULL, *py_t=NULL, *py_w=NULL;
+    fitpack::d_ssize_t nc;
+    int k;   // NB: declare as npy_intp, and it's garbage
```
That's a note to self I forgot to clear up, thanks for spotting it! Will remove.
(The note was about the fact that `PyArg_ParseTuple below is sensitive to integer types/formats: "n" for Py_ssize_t and "i" for int; a mismatch make PyArg_ParseTuple silently return garbage).

ev-br committed:
MAINT: interpolate: use int64_t directly w/o typedef-ing

ev-br committed:
MAINT: interpolate: Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ev-br commented:
This is ready for review. 

This now removes most of the _bspl.pyx Cython extension to manually wrapped C++. With this, we halve the binary size of `_bspl.so` vs `_dierckx.so` which this PR introduces.

Several TODO items are IMO best left for a follow-up:
- two remaining Cython-implemented functions will follow the same  pattern; lumping them into this PR is just increasing the PR size and does not bring anything fundamentally new.
- several small style/perf comments are down to how to zero out a buffer: element-by-element in a loop or separately. This will need measuring and is surely can be done in a follow-up.

The basic design is there, and the key question IMO is whether we add footguns by ditching Cython in favor of a manually written C extension. It'd be great to see if somebody sees problems I missed.
An earlier version of this was reviewed by @ngoldbaum (thanks Nathan!).

Kai-Striega commented:
@ev-br I'm back from my unplanned hiatus. I'm happy to do it. I'm already reviewing a different PR today, I'll try to get to it this fortnight, hopefully next weekend.

Kai-Striega commented on scipy/interpolate/src/_dierckxmodule.cc:
```diff
@@ -0,0 +1,775 @@
+};
+
+template <typename T>
+using unique_fptr = std::unique_ptr<T, malloc_deleter>;
```
You're using this to wrap "c-style" arrays. I think it would be preferable to use ``std::array``/``std::vector`` instead. This will remove the need for this template entirely, as ``std::unique_ptr`` calls ``T``'s destructor by default.

ev-br replied:
Makes sense! 
Pushed a commit where `wrk` arrays are allocated as `vector<double> wrk; wrk.resize(2*k+2)` instead of this, admittedly, fairly confusing incantation.

Kai-Striega reviewed (commented):
Had a look through, a couple of inline comments.

My main issue is with how we've defined _another_ array type to use within SciPy. Maybe we could investigate using ``kokkos::mdspan`` instead?

ev-br committed:
MAINT: interpolate: address code review comments

use vectors not unique_ptr(malloc) for work arrays

ev-br committed:
MAINT: interpolate/_dierckxmodule: flip bounds checking back to false

ev-br commented on scipy/interpolate/src/_dierckxmodule.cc:
```diff
@@ -0,0 +1,775 @@
+ * RAII malloc-d work arrays
+ */
+struct malloc_deleter {
+    void operator()(void *p) const { std::free(p); }
```
I know. There is an internal function, `_de_Boor_D`, which expects a raw pointer to pre-allocated arena, and this was a way to `malloc` the space and have it deallocated automagically. Of course, this is a recipe from somewhere in the internet :-).

Anyhow, this is now gone in favor of just creating a `vector` in a function scope. The sequence is 

```
{
   std::vector<double> wrk;
   wrk.resize(2*k+2);

   call_low_level_stuff(...., wrk.data());

  ...

}  // wrk goes out of scope and is free-d
```

Kai-Striega replied:
Your example can be simplified using constructor (4) from [cppreference/vector](https://en.cppreference.com/w/cpp/container/vector/vector).

```cpp
{
   std::vector<double> wrk(2*k+2);

   call_low_level_stuff(...., wrk.data());

  ...

}  // wrk goes out of scope and is free-d
```

This will create an array of size ``2*k + 2`` filled with whatever [std::allocator_traits](http://en.cppreference.com/w/cpp/memory/allocator_traits)<A>::construct(m, p); returns for ``double``

ev-br commented on scipy/interpolate/src/_dierckxmodule.cc:
```diff
@@ -0,0 +1,775 @@
+    // heavy lifting happens here
+    try {
+        double new_knot = fitpack::fpknot(
+            static_cast<const double *>(PyArray_DATA(a_x)), PyArray_DIM(a_x, 0),
+            static_cast<const double *>(PyArray_DATA(a_t)), PyArray_DIM(a_t, 0),
```
Hm, it's an implicit type conversion? `npy_intp` is either int32 or int64, depending on who knows what; if it's int32, the automatic conversion to int64 is safe, unless I'm missing something?

Kai-Striega replied:
I don't think you are. I'm just not sure how it works as well and thought I'd point it out.

Yes int32 -> int64  should be safe.

ev-br commented:
Thanks Kai for the review!

I completely agree that `Array1D` / `Array2D` is very much a temporary implement, with the only purpose being to toggle boundschecking. And it's the dealbreaker IMO.

I'be happy to see the back of `Array2D` and change its usage into a more standard way. Preferably in a follow-up PR though, this one is large enough.

We had a discussion about this a while ago with @izaid who gave me very helpful hints (the key observation is that boundschecking for mdspan needs a LayoutPolicy), and here's a somewhat outdated branch with experiments
https://github.com/ev-br/scipy/compare/splrep_repro...ev-br:scipy:splrep_repro_with_mdspan2?expand=1  

Would you be interested in playing with mdspan instead of these clumsy `Array{1D,2D}`? This would be great. Otherwise I'll hopefully get to it at some point (tm).

Kai-Striega commented:
> Thanks Kai for the review!

You're welcome. It's great and difficult work!

> I completely agree that Array1D / Array2D is very much a temporary implement, with the only purpose being to toggle boundschecking. And it's the dealbreaker IMO.
>
> I'be happy to see the back of Array2D and change its usage into a more standard way. Preferably in a follow-up PR though, this one is large enough.

I agree that this should be temporary. As this PR doesn't introduce them, I don't think it should block this PR from going into the main branch. We should probably create an issue to keep track of the rework.

> We had a discussion about this a while ago with @izaid who gave me very helpful hints (the key observation is that boundschecking for mdspan needs a LayoutPolicy), and here's a somewhat outdated branch with experiments
https://github.com/ev-br/scipy/compare/splrep_repro...ev-br:scipy:splrep_repro_with_mdspan2?expand=1
>
> Would you be interested in playing with mdspan instead of these clumsy Array{1D,2D}? This would be great. Otherwise I'll hopefully get to it at some point (tm).

I'm going to allocate some time to play around with it. However, I've already allocated a fair bit of time for my holiday break to code/open source. I'm hesitant to overcommit.

ev-br commented:
All review comments are addressed.

Kai-Striega commented:
@ev-br are we waiting for anything else here?

ev-br commented:
> are we waiting for anything else here?

Not to my knowleege, no.

Kai-Striega commented:
I'm not qualified to assess this PR from a maths perspective, but, as fair as I can tell, it looks good from a technical perspective. If you need me to I could spend some time reading up on the maths and merge it.

ev-br commented:
Thanks Kai! This PR is pretty much purely technical, there is no new math really, just moving things from Cython to either pure python or C++ / python C API. So it's up to you as a reviewer I suppose whether you want to assess the maths along with the technical content. In my mind the main part of this patch is moving from Cython's doing quite a bit of handholding with the C API to a more manual approach.

Kai-Striega commented:
@ev-br I'm happy with this. However, given how complex the changes are technically, and the fact that I'm not a C++ expert, I'm going to leave this open for another week or so to see if anyone objects. Otherwise, I'll merge it.

Kai-Striega commented:
It's been 3 weeks and no-one has raised any objections. I'm going to merge this. Thanks for your work @ev-br!
